[11.0] backport sale_timesheet_rounded from 12.0

grindtildeath opened the pull request:
Backport of #257

simahawk committed:
Add hr_timesheet_rounded

grindtildeath committed:
Improvements

Small improvements

Improve tests
Add missing view
Copy fields lists
Do not obfuscate fields import

Semantics

Improve doc

Rename module to sale_timesheet_rounded

Fix no rounding when explicitely setting 0 as rounded amount

Allow no rounding on projects

Although there was a test setting the values to False, this was impossible
on the UI because of the required on timesheet_rounding_method and no
proper selection.

Initialize unit_amount_rounded with the value of unit_amount

grindtildeath committed:
Fix design issue and always set unit_amount_rounded

The previous implementation relied on the override of _timesheet_postprocess
to compute unit_amount_rounded only if the line fulfils the condition that
were defined in _update_amount_rounded.
The issue is that this was done only for analytic line having a project_id
and the value stayed 0.0 for any other line.
However, as the qty_delivered on sale order line is computed using a read_group
we cannot ensure all the read lines will have a project_id, so if we replace
the value of unit_amount by unit_amount_rounded, we must make sure the value
is set even if the line is not linked to a project with a rounding policy.
Therefore unit_amount_rounded is now always updated, and is added to the
triggering fields of sale.order.line._compute_qty_delivered to make sure the
computation is correct.

Remove the check on the product expense policy

Because expense lines are lines without project_id and it is already checked

Do not copy the value of unit_amount_rounded to make sure it recomputed

Update README

grindtildeath committed:
[MIG] sale_timesheet_rounded: Backport from 12.0

Fix labels on the views

grindtildeath committed:
project_task_stage_allow_timesheet: Do not break tests of other modules

i-vyshnevska commented on project_task_stage_allow_timesheet/models/account_analytic_line.py:
```diff
@@ -15,6 +16,11 @@ class AccountAnalyticLine(models.Model):
+            config['test_enable'] and
+            not self.env.context.get('test_task_stage_allow_timesheet')
+        ):
+            return
```
this is weird tests shouldn't imply on  logic in method

i-vyshnevska commented on project_task_stage_allow_timesheet/tests/test_project_task_stage_allow_timesheet.py:
```diff
@@ -10,6 +10,10 @@ class TestProjectTaskStageAllowTimesheet(SavepointCase):
+        cls.env = cls.env(context=dict(
+            cls.env.context, tracking_disable=True,
+            test_task_stage_allow_timesheet=True
+        ))
```
maybe additional flags needs to force tests

i-vyshnevska commented on sale_timesheet_rounded/models/account_analytic_line.py:
```diff
@@ -0,0 +1,167 @@
+    def _calc_unit_amount_rounded(self):
+        self.ensure_one()
+        force_compute = self.env.context.get('force_compute')
+        # TODO: do we still need this?
```
or more explicit comment or remove it

simahawk committed:
sale_timesheet_rounded: fix do not use default

leemannd reviewed (approved):
Tested functionally

leemannd commented:
@simahawk @alexey-pelykh If you have time to take a look at it. It's worth merging

alexey-pelykh reviewed (commented):
Why commit history is so odd, if sale_timesheet_rounded was added via 1 commit in 12?

leemannd commented:
@alexey-pelykh You're right about the commit history.
@grindtildeath do you think you can update this?

grindtildeath commented:
@alexey-pelykh @leemannd Actually, the module wasn't added via 1 commit in 12... At the beginning were the first 6 commits from @simahawk that you see here. Then I fixed the issues with all of the commits you see here on 11.0, which have been forward ported into #257 (https://github.com/simahawk/timesheet/pulls?q=is%3Apr+is%3Aclosed) before being squashed (without taking care of the author) into the same one, after the last commit added by @simahawk here...

So this commit history was actually "the right one", but now I squashed by authors and kept important commits messages so that we can merge...

leemannd commented:
@grindtildeath Thanks! :+1:
